qualtran/bloqs/data_loading/select_swap_qrom.py

```diff
@@ -239,81 +246,142 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
-            for reg in qrom.target_registers
-        }
-        qrom_op = qrom.on_registers(**qrom_ctrls, **qrom_selection, **qrom_targets)
-        # 2. Construct SwapWithZero.
-        swz_ops = []
-        assert len(qrom_targets) == len(self.swap_with_zero_bloqs)
-        for targets, swz in zip(qrom_targets.values(), self.swap_with_zero_bloqs):
-            if len(targets) <= 1:
-                continue
-            swz_selection = {
-                swz_reg.name: quregs[sel_reg.name][-swz_reg.bitsize :]
-                for sel_reg, swz_reg in zip(self.selection_registers, swz.selection_registers)
-            }
-            swz_ops.append(swz.on_registers(**swz_selection, targets=targets))
-        # 3. Construct CNOTs from 0th borrowed register to clean target registers.
-        cnot_ops = []
-        for qrom_reg, target_reg in zip(qrom.target_registers, self.target_registers):
-            qrom_batched_target = qrom_targets[qrom_reg.name]
+        bb: 'BloqBuilder',
+        ctrls: List['SoquetT'],
+        sel_l: List['SoquetT'],
+        targets: List['SoquetT'],
+        uncompute: bool = False,
+    ) -> Tuple[List['SoquetT'], List['SoquetT'], List['SoquetT']]:
+        in_soqs = {reg.name: soq for reg, soq in zip(self.qrom_bloq.control_registers, ctrls)}
+        in_soqs |= {reg.name: soq for reg, soq in zip(self.qrom_bloq.selection_registers, sel_l)}
+        in_soqs |= {reg.name: soq for reg, soq in zip(self.qrom_bloq.target_registers, targets)}
+        out_soqs = bb.add_d(self.qrom_bloq.adjoint() if uncompute else self.qrom_bloq, **in_soqs)
+        ctrls = [out_soqs[reg.name] for reg in self.qrom_bloq.control_registers]
+        sel_l = [out_soqs[reg.name] for reg in self.qrom_bloq.selection_registers]
+        targets = [out_soqs[reg.name] for reg in self.qrom_bloq.target_registers]
+        return ctrls, sel_l, targets
+
+    def _add_swap_with_zero_bloq(
+        self,
+        bb: 'BloqBuilder',
+        selection: List['SoquetT'],
+        targets: List['SoquetT'],
+        uncompute: bool = False,
+    ) -> Tuple[List['SoquetT'], List['SoquetT']]:
+        # Get soquets for SwapWithZero
+        assert len(targets) == len(self.swap_with_zero_bloqs)
+        sel_names = [reg.name for reg in self.swap_with_zero_bloqs[0].selection_registers]
+        soqs = {sel_name: soq for sel_name, soq in zip(sel_names, selection)}
+        out_targets: List['SoquetT'] = []
+        for target, swz in zip(targets, self.swap_with_zero_bloqs):
+            soqs['targets'] = target
+            soqs = bb.add_d(swz.adjoint() if uncompute else swz, **soqs)
+            out_targets.append(soqs['targets'])
+        return [soqs[reg_name] for reg_name in sel_names], out_targets
+
+    def _add_cnot(
+        self, bb: 'BloqBuilder', qrom_targets: List['SoquetT'], target: List['SoquetT']
+    ) -> Tuple[List['SoquetT'], List['SoquetT']]:
+        for i, qrom_reg in enumerate(qrom_targets):
+            assert isinstance(qrom_reg, np.ndarray)
```

Comment:
Add an error message or remove assert if just for debugging.

Reply:
This is to make mypy happy since I access `qrom_reg.shape` below. I'll add a comment.